src/dmd/mars.d

```diff
@@ -769,12 +749,72 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
             }
         }
     }
+
+    // Output the makefile dependencies
+    if (params.emitMakeDeps)
+        emitMakeDeps(params, library);
+
     if (global.errors || global.warnings)
         removeHdrFilesAndFail(params, modules);
 
     return status;
 }
 
+/// Emit the makefile dependencies for the -makedeps switch
+version (NoMain) {} else
+{
+    void emitMakeDeps(ref Param params, Library library)
+    {
+        assert(params.emitMakeDeps);
+
+        OutBuffer buf;
+
+        // start by resolving and writing the target (which is sometimes resolved during link phase)
+        if (params.link && params.exefile)
+        {
+            buf.writeEscapedMakePath(&params.exefile[0]);
+        }
+        else if (params.lib && library)
+        {
+            buf.writeEscapedMakePath(library.getFilename());
+        }
+        else if (params.objname)
+        {
+            buf.writeEscapedMakePath(&params.objname[0]);
+        }
+        else if (params.objfiles.length)
+        {
+            buf.writeEscapedMakePath(params.objfiles[0]);
+            foreach (of; params.objfiles[1 .. $])
+            {
+                buf.writestring(" ");
+                buf.writeEscapedMakePath(of);
+            }
```

Comment:
So IIUC, *every* object file is invalidated as soon as one of the dependencies changes?

Reply:
If the makefile is generated by a single invocation creating multiple object files from multiple source files, yes.
It is up to the build system to define granularity of invocation.
Build systems that read such dependencies emitted by the compiler will typically want to have one invocation per source file.
It makes initial build longer, but edit-compile-run cycles are much shorter.

Reply:
I was thinking about the best case here for incremental debug builds - the build system initially starting with an all-at-once or per-D-package compilation, and later only recompiling the invalidated subset of object files using the same granularity (all-at-once/per-package). - No idea if something like this is supported by any existing build system...

Reply:
I don't think this will work with the current template emission strategy, the build system can't predict in which object files symbols are written to.

Reply:
If the build is done per package or per module, then there should be one object file per package or per module, respectively.

Reply:
Using separate object files and rebuilding only the invalidated subset should be definitely faster than using a single object file if the code isn't template-heavy. Intelligently adapting the root modules provided in the cmdline is probably not feasible for mainstream build systems though.

> I don't think this will work with the current template emission strategy

Very likely; not sure if current `-allinst` semantics would suffice to avoid that problem. It should work with LDC's upcoming `-linkonce-templates`.

Reply:
`-allinst` *should* be fixed by now. I've been building separately for years now and the only problems I've had with templates are problems that aren't solved building all-at-once.